be/src/types/bitmap_value.cpp

```diff
@@ -923,91 +923,198 @@ void BitmapValue::_convert_to_smaller_type() {
 }
 
 int64_t BitmapValue::sub_bitmap_internal(const int64_t& offset, const int64_t& len, BitmapValue* ret_bitmap) {
-    if (offset > 0 && offset >= _bitmap->cardinality()) {
-        return 0;
-    }
-    if (offset < 0 && std::abs(offset) > _bitmap->cardinality()) {
+    switch (_type) {
+    case EMPTY:
         return 0;
+    case SINGLE: {
+        if (offset >= 1 || offset < -1 || len <= 0) {
+            return 0;
+        } else {
+            ret_bitmap->add(_sv);
+            return 1;
+        }
     }
-    int64_t abs_offset = offset;
-    if (offset < 0) {
-        abs_offset = _bitmap->cardinality() + offset;
-    }
+    case SET: {
+        if (offset > 0 && offset >= _set->size()) {
+            return 0;
+        }
+        if (offset < 0 && std::abs(offset) > _set->size()) {
+            return 0;
+        }
+        int64_t abs_offset = offset;
+        if (offset < 0) {
+            abs_offset = _set->size() + offset;
```

Comment:
why add, must abs_offset > 0?

Reply:
yep. this is original codes. I just move it here.